add index for partition table find next partition ID failed

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. after use ctrl-c cancel reorganize partition as describe in #42447 
2. add index for the table 
```sql
alter table t1 add index idx_kc (k, c);
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
add index success 
### 3. What did you see instead (Required)
1. add failed
```sql
TiDB root@127.0.0.1:db1> alter table t1 add index idx_kc (k, c);
You're about to run a destructive command.
Do you want to proceed? (y/n): y
Your call!
(1105, 'DDL job rollback, error msg: partition id not found 3898')
```
3. log
```log
[2023/03/21 19:02:41.377 +08:00] [INFO] [ddl.go:1068] ["[ddl] start DDL job"] [job="ID:4391, Type:add index, State:queueing, SchemaState:none, SchemaID:324, TableID:3897, RowCount:0, ArgLen:6, start time: 2023-03-21 19:02:41.321 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0, UniqueWarnings:0"] [query="alter table t1 add index idx_kc (k, c)"]
[2023/03/21 19:02:41.377 +08:00] [INFO] [ddl_worker.go:234] ["[ddl] add DDL jobs"] ["batch count"=1] [jobs="ID:4391, Type:add index, State:queueing, SchemaState:none, SchemaID:324, TableID:3897, RowCount:0, ArgLen:6, start time: 2023-03-21 19:02:41.321 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0, UniqueWarnings:0; "] [table=true]
[2023/03/21 19:02:41.397 +08:00] [INFO] [ddl_worker.go:960] ["[ddl] run DDL job"] [worker="worker 5, tp add index"] [job="ID:4391, Type:add index, State:queueing, SchemaState:none, SchemaID:324, TableID:3897, RowCount:0, ArgLen:0, start time: 2023-03-21 19:02:41.321 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0, UniqueWarnings:0"]
[2023/03/21 19:02:41.400 +08:00] [INFO] [index.go:613] ["[ddl] run add index job"] [job="ID:4391, Type:add index, State:running, SchemaState:none, SchemaID:324, TableID:3897, RowCount:0, ArgLen:6, start time: 2023-03-21 19:02:41.321 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0, UniqueWarnings:0"] [indexInfo="{\"id\":2,\"idx_name\":{\"O\":\"idx_kc\",\"L\":\"idx_kc\"},\"tbl_name\":{\"O\":\"\",\"L\":\"\"},\"idx_cols\":[{\"name\":{\"O\":\"k\",\"L\":\"k\"},\"offset\":1,\"length\":-1},{\"name\":{\"O\":\"c\",\"L\":\"c\"},\"offset\":2,\"length\":-1}],\"state\":0,\"backfill_state\":0,\"comment\":\"\",\"index_type\":1,\"is_unique\":false,\"is_primary\":false,\"is_invisible\":false,\"is_global\":false,\"mv_index\":false}"]
[2023/03/21 19:02:41.419 +08:00] [INFO] [domain.go:224] ["diff load InfoSchema success"] [currentSchemaVersion=294] [neededSchemaVersion=295] ["start time"=3.02062ms] [gotSchemaVersion=295] [phyTblIDs="[3897,3898,3899,3900,3901,3902,3903]"] [actionTypes="[7,7,7,7,7,7,7]"]
[2023/03/21 19:02:41.422 +08:00] [INFO] [domain.go:795] ["mdl gets lock, update to owner"] [jobID=4391] [version=295]
.....


[2023/03/21 19:02:46.644 +08:00] [INFO] [index_cop.go:79] ["[ddl-ingest] finish a cop-request task"] [id=4]
[2023/03/21 19:02:46.644 +08:00] [INFO] [index_cop.go:84] ["[ddl-ingest] task is found in results"] [id=4] [task="taskID: 4, physicalTableID: 3898, range: [748000000000000f3a5f72800000000016e392, 748000000000000f3a5f7280000000001e847f], jobID: 4391"]
[2023/03/21 19:02:46.644 +08:00] [INFO] [backfilling.go:376] ["[ddl] backfill worker finish task"] [worker="backfill-worker 0, tp ingest index"] [task="taskID: 4, physicalTableID: 3898, range: [748000000000000f3a5f72800000000016e392, 748000000000000f3a5f7280000000001e847f], jobID: 4391"] ["added count"=771310] ["scan count"=771310] ["next key"=748000000000000f3a5f7280000000001e847f] ["take time"=3.80479982s]
[2023/03/21 19:02:46.653 +08:00] [INFO] [backfilling.go:617] ["[ddl] backfill workers successfully processed batch"] [element=ID:2,TypeKey:_idx_] ["total added count"=1999999] ["start key"=748000000000000f3a5f728000000000000001] ["next key"=748000000000000f3a5f7280000000001e847f] ["batch added count"=1999999] ["take time"=4.492926233s] []
[2023/03/21 19:02:46.653 +08:00] [INFO] [index_cop.go:217] ["[ddl-ingest] close cop-request sender pool"] ["results not handled"=0]
[2023/03/21 19:02:46.654 +08:00] [INFO] [backfilling.go:460] ["[ddl] backfill worker exit"] [worker="backfill-worker 0, tp ingest index"]
[2023/03/21 19:02:46.654 +08:00] [INFO] [backfilling.go:460] ["[ddl] backfill worker exit"] [worker="backfill-worker 2, tp ingest index"]
[2023/03/21 19:02:46.654 +08:00] [INFO] [backfilling.go:460] ["[ddl] backfill worker exit"] [worker="backfill-worker 1, tp ingest index"]
[2023/03/21 19:02:46.654 +08:00] [INFO] [backfilling.go:460] ["[ddl] backfill worker exit"] [worker="backfill-worker 3, tp ingest index"]
[2023/03/21 19:02:46.654 +08:00] [ERROR] [index.go:1972] ["[ddl] find next partition ID failed"] [table="{\"Columns\":[{\"id\":1,\"name\":{\"O\":\"id\",\"L\":\"id\"},\"offset\":0,\"origin_default\":null,\"origin_default_bit\":null,\"default\":null,\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":8,\"Flag\":515,\"Flen\":20,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":2,\"name\":{\"O\":\"k\",\"L\":\"k\"},\"offset\":1,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"0\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":3,\"Flag\":9,\"Flen\":11,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":3,\"name\":{\"O\":\"c\",\"L\":\"c\"},\"offset\":2,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":120,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":4,\"name\":{\"O\":\"pad\",\"L\":\"pad\"},\"offset\":3,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":60,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null}],\"PublicColumns\":[{\"id\":1,\"name\":{\"O\":\"id\",\"L\":\"id\"},\"offset\":0,\"origin_default\":null,\"origin_default_bit\":null,\"default\":null,\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":8,\"Flag\":515,\"Flen\":20,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":2,\"name\":{\"O\":\"k\",\"L\":\"k\"},\"offset\":1,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"0\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":3,\"Flag\":9,\"Flen\":11,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":3,\"name\":{\"O\":\"c\",\"L\":\"c\"},\"offset\":2,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":120,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":4,\"name\":{\"O\":\"pad\",\"L\":\"pad\"},\"offset\":3,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":60,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null}],\"VisibleColumns\":[{\"id\":1,\"name\":{\"O\":\"id\",\"L\":\"id\"},\"offset\":0,\"origin_default\":null,\"origin_default_bit\":null,\"default\":null,\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":8,\"Flag\":515,\"Flen\":20,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":2,\"name\":{\"O\":\"k\",\"L\":\"k\"},\"offset\":1,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"0\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":3,\"Flag\":9,\"Flen\":11,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":3,\"name\":{\"O\":\"c\",\"L\":\"c\"},\"offset\":2,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":120,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":4,\"name\":{\"O\":\"pad\",\"L\":\"pad\"},\"offset\":3,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":60,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null}],\"HiddenColumns\":[],\"WritableColumns\":[{\"id\":1,\"name\":{\"O\":\"id\",\"L\":\"id\"},\"offset\":0,\"origin_default\":null,\"origin_default_bit\":null,\"default\":null,\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":8,\"Flag\":515,\"Flen\":20,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":2,\"name\":{\"O\":\"k\",\"L\":\"k\"},\"offset\":1,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"0\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":3,\"Flag\":9,\"Flen\":11,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":3,\"name\":{\"O\":\"c\",\"L\":\"c\"},\"offset\":2,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":120,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":4,\"name\":{\"O\":\"pad\",\"L\":\"pad\"},\"offset\":3,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":60,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null}],\"FullHiddenColsAndVisibleColumns\":[{\"id\":1,\"name\":{\"O\":\"id\",\"L\":\"id\"},\"offset\":0,\"origin_default\":null,\"origin_default_bit\":null,\"default\":null,\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":8,\"Flag\":515,\"Flen\":20,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":2,\"name\":{\"O\":\"k\",\"L\":\"k\"},\"offset\":1,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"0\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":3,\"Flag\":9,\"Flen\":11,\"Decimal\":0,\"Charset\":\"binary\",\"Collate\":\"binary\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":3,\"name\":{\"O\":\"c\",\"L\":\"c\"},\"offset\":2,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":120,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null},{\"id\":4,\"name\":{\"O\":\"pad\",\"L\":\"pad\"},\"offset\":3,\"origin_default\":null,\"origin_default_bit\":null,\"default\":\"\",\"default_bit\":null,\"default_is_expr\":false,\"generated_expr_string\":\"\",\"generated_stored\":false,\"dependences\":null,\"type\":{\"Tp\":254,\"Flag\":1,\"Flen\":60,\"Decimal\":0,\"Charset\":\"utf8mb4\",\"Collate\":\"utf8mb4_bin\",\"Elems\":null,\"ElemsIsBinaryLit\":null,\"Array\":false},\"state\":5,\"comment\":\"\",\"hidden\":false,\"change_state_info\":null,\"version\":2,\"GeneratedExpr\":null,\"DefaultExpr\":null}]}"] [error="partition id not found 3898"]
[2023/03/21 19:02:46.655 +08:00] [WARN] [reorg.go:234] ["[ddl] run reorg job done"] ["handled rows"=1999999] [error="partition id not found 3898"]
[2023/03/21 19:02:46.655 +08:00] [INFO] [backend.go:469] ["engine close start"] [engineTag=t1:2] [engineUUID=a0f82a3c-5421-5636-a09c-8e48474d5839]
```
### 4. What is your TiDB version? (Required)
```sql
+-----------------------------------------------------------+
| tidb_version()                                            |
+-----------------------------------------------------------+
| Release Version: v7.0.0                                   |
| Edition: Community                                        |
| Git Commit Hash: fc166ffae8576ac3efb2f5abac848d5edbaf2657 |
| Git Branch: heads/refs/tags/v7.0.0                        |
| UTC Build Time: 2023-03-20 07:08:59                       |
| GoVersion: go1.20.2                                       |
| Race Enabled: false                                       |
| TiKV Min Version: 6.2.0-alpha                             |
| Check Table Before Drop: false                            |
| Store: tikv                                               |
+-----------------------------------------------------------+

```
<!-- Paste the output of SELECT tidb_version() -->

--- a/ddl/partition.go
+++ b/ddl/partition.go
@@ -1741,7 +1741,9 @@ func (w *worker) onDropTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (
 		return ver, errors.Trace(err)
 	}
 	if job.Type == model.ActionAddTablePartition || job.Type == model.ActionReorganizePartition {
-		// It is rollbacked from adding table partition, just remove addingDefinitions from tableInfo.
+		// It is rollback from reorganize partition, just remove DroppingDefinitions from tableInfo
+		tblInfo.Partition.DroppingDefinitions = nil
+		// It is rollback from adding table partition, just remove addingDefinitions from tableInfo.
 		physicalTableIDs, pNames, rollbackBundles := rollbackAddingPartitionInfo(tblInfo)
 		err = infosync.PutRuleBundlesWithDefaultRetry(context.TODO(), rollbackBundles)
 		if err != nil {

--- a/ddl/partition_test.go
+++ b/ddl/partition_test.go
@@ -16,7 +16,9 @@ package ddl_test
 
 import (
 	"testing"
+	"time"
 
+	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/ddl"
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/parser/model"
@@ -138,4 +140,101 @@ func testTruncatePartition(t *testing.T, ctx sessionctx.Context, d ddl.DDL, dbIn
 	v := getSchemaVer(t, ctx)
 	checkHistoryJobArgs(t, ctx, job.ID, &historyJobArgs{ver: v, tbl: tblInfo})
 	return job
+}
+
+func TestReorganizePartitionRollback(t *testing.T) {
+	// See issue: https://github.com/pingcap/tidb/issues/42448
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/reorgPartWriteReorgReplacedPartIDsFail", "return(true)"))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/ddl/reorgPartWriteReorgReplacedPartIDsFail"))
+	}()
+
+	store, do := testkit.CreateMockStoreAndDomain(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("CREATE TABLE `t1` (\n" +
+		"    `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+		"    `k` int(11) NOT NULL DEFAULT '0',\n" +
+		"    `c` char(120) NOT NULL DEFAULT '',\n" +
+		"    `pad` char(60) NOT NULL DEFAULT '',\n" +
+		"    PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,\n" +
+		"    KEY `k_1` (`k`)\n" +
+		"  ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin\n" +
+		"  PARTITION BY RANGE (`id`)\n" +
+		"  (PARTITION `p0` VALUES LESS THAN (2000000),\n" +
+		"   PARTITION `p1` VALUES LESS THAN (4000000),\n" +
+		"   PARTITION `p2` VALUES LESS THAN (6000000),\n" +
+		"   PARTITION `p3` VALUES LESS THAN (8000000),\n" +
+		"   PARTITION `p4` VALUES LESS THAN (10000000),\n" +
+		"   PARTITION `p5` VALUES LESS THAN (MAXVALUE))")
+	tk.MustExec("insert into t1(k, c, pad) values (1, 'a', 'beijing'), (2, 'b', 'chengdu')")
+	ddlDone := make(chan error)
+	defer close(ddlDone)
+	go func() {
+		tk2 := testkit.NewTestKit(t, store)
+		tk2.MustExec("use test")
+		err := tk2.ExecToErr("alter table t1 reorganize partition p0, p1, p2, p3, p4 into( partition pnew values less than (10000000))")
+		ddlDone <- err
+	}()
+
+	showDDLJobStart := time.Now()
+	jobID := ""
+	schemaState := ""
+	for time.Since(showDDLJobStart) < time.Minute {
+		rows := tk.MustQuery("admin show ddl jobs where JOB_TYPE='alter table reorganize partition'").Rows()
+		if len(rows) == 0 {
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+
+		require.Equal(t, 1, len(rows))
+		jobID = rows[0][0].(string)
+		schemaState = rows[0][4].(string)
+		if schemaState != "write reorganization" {
+			time.Sleep(500 * time.Millisecond)
+			continue
+		}
+		break
+	}
+
+	require.NotEqual(t, "", jobID)
+	require.Equal(t, "write reorganization", schemaState)
+	tk.MustExec("admin cancel ddl jobs " + jobID)
+
+	select {
+	case err := <-ddlDone:
+		require.Error(t, err)
+	case <-time.After(time.Minute):
+		require.FailNow(t, "wait ddl cancelled timeout")
+	}
+
+	// check job rollback finished
+	rows := tk.MustQuery("admin show ddl jobs where JOB_ID=" + jobID).Rows()
+	require.Equal(t, 1, len(rows))
+	require.Equal(t, "rollback done", rows[0][len(rows[0])-1])
+
+	// check table meta after rollback
+	tk.MustQuery("show create table t1").Check(testkit.Rows("t1 CREATE TABLE `t1` (\n" +
+		"  `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+		"  `k` int(11) NOT NULL DEFAULT '0',\n" +
+		"  `c` char(120) NOT NULL DEFAULT '',\n" +
+		"  `pad` char(60) NOT NULL DEFAULT '',\n" +
+		"  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,\n" +
+		"  KEY `k_1` (`k`)\n" +
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=5001\n" +
+		"PARTITION BY RANGE (`id`)\n" +
+		"(PARTITION `p0` VALUES LESS THAN (2000000),\n" +
+		" PARTITION `p1` VALUES LESS THAN (4000000),\n" +
+		" PARTITION `p2` VALUES LESS THAN (6000000),\n" +
+		" PARTITION `p3` VALUES LESS THAN (8000000),\n" +
+		" PARTITION `p4` VALUES LESS THAN (10000000),\n" +
+		" PARTITION `p5` VALUES LESS THAN (MAXVALUE))"))
+	tbl, err := do.InfoSchema().TableByName(model.NewCIStr("test"), model.NewCIStr("t1"))
+	require.NoError(t, err)
+	require.NotNil(t, tbl.Meta().Partition)
+	require.Nil(t, tbl.Meta().Partition.AddingDefinitions)
+	require.Nil(t, tbl.Meta().Partition.DroppingDefinitions)
+
+	// test then add index should success
+	tk.MustExec("alter table t1 add index idx_kc (k, c)")
 }

--- a/ddl/partition_test.go
+++ b/ddl/partition_test.go
@@ -18,8 +18,8 @@ import (
 	"testing"
 	"time"
 
-	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/ddl"
+	"github.com/pingcap/tidb/ddl/internal/callback"
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/parser/model"
 	"github.com/pingcap/tidb/parser/mysql"
@@ -144,11 +144,6 @@ func testTruncatePartition(t *testing.T, ctx sessionctx.Context, d ddl.DDL, dbIn
 
 func TestReorganizePartitionRollback(t *testing.T) {
 	// See issue: https://github.com/pingcap/tidb/issues/42448
-	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/reorgPartWriteReorgReplacedPartIDsFail", "return(true)"))
-	defer func() {
-		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/ddl/reorgPartWriteReorgReplacedPartIDsFail"))
-	}()
-
 	store, do := testkit.CreateMockStoreAndDomain(t)
 	tk := testkit.NewTestKit(t, store)
 	tk.MustExec("use test")
@@ -168,8 +163,20 @@ func TestReorganizePartitionRollback(t *testing.T) {
 		"   PARTITION `p4` VALUES LESS THAN (10000000),\n" +
 		"   PARTITION `p5` VALUES LESS THAN (MAXVALUE))")
 	tk.MustExec("insert into t1(k, c, pad) values (1, 'a', 'beijing'), (2, 'b', 'chengdu')")
+
+	wait := make(chan struct{})
+	defer close(wait)
 	ddlDone := make(chan error)
 	defer close(ddlDone)
+	hook := &callback.TestDDLCallback{Do: do}
+	hook.OnJobRunAfterExported = func(job *model.Job) {
+		if job.Type == model.ActionReorganizePartition && job.SchemaState == model.StateWriteReorganization {
+			<-wait
+			<-wait
+		}
+	}
+	do.DDL().SetHook(hook)
+
 	go func() {
 		tk2 := testkit.NewTestKit(t, store)
 		tk2.MustExec("use test")
@@ -177,30 +184,27 @@ func TestReorganizePartitionRollback(t *testing.T) {
 		ddlDone <- err
 	}()
 
-	showDDLJobStart := time.Now()
 	jobID := ""
-	schemaState := ""
-	for time.Since(showDDLJobStart) < time.Minute {
-		rows := tk.MustQuery("admin show ddl jobs where JOB_TYPE='alter table reorganize partition'").Rows()
-		if len(rows) == 0 {
-			time.Sleep(100 * time.Millisecond)
-			continue
-		}
 
+	// wait DDL job reaches hook and then cancel
+	select {
+	case wait <- struct{}{}:
+		rows := tk.MustQuery("admin show ddl jobs where JOB_TYPE='alter table reorganize partition'").Rows()
 		require.Equal(t, 1, len(rows))
 		jobID = rows[0][0].(string)
-		schemaState = rows[0][4].(string)
-		if schemaState != "write reorganization" {
-			time.Sleep(500 * time.Millisecond)
-			continue
-		}
-		break
+		tk.MustExec("admin cancel ddl jobs " + jobID)
+	case <-time.After(time.Minute):
+		require.FailNow(t, "timeout")
 	}
 
-	require.NotEqual(t, "", jobID)
-	require.Equal(t, "write reorganization", schemaState)
-	tk.MustExec("admin cancel ddl jobs " + jobID)
+	// continue to run DDL
+	select {
+	case wait <- struct{}{}:
+	case <-time.After(time.Minute):
+		require.FailNow(t, "timeout")
+	}
 
+	// wait ddl done
 	select {
 	case err := <-ddlDone:
 		require.Error(t, err)

--- a/ddl/rollingback.go
+++ b/ddl/rollingback.go
@@ -355,6 +355,21 @@ func rollingbackTruncateTable(t *meta.Meta, job *model.Job) (ver int64, err erro
 	return cancelOnlyNotHandledJob(job, model.StateNone)
 }
 
+func rollingbackReorganizePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	if job.SchemaState == model.StateNone {
+		job.State = model.JobStateCancelled
+		return ver, dbterror.ErrCancelledDDLJob
+	}
+
+	tblInfo, err := GetTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// addingDefinitions is also in tblInfo, here pass the tblInfo as parameter directly.
+	return convertAddTablePartitionJob2RollbackJob(d, t, job, dbterror.ErrCancelledDDLJob, tblInfo)
+}
+
 func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
 	switch job.Type {
 	case model.ActionAddColumn:
@@ -365,6 +380,8 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job) 
 		ver, err = rollingbackAddIndex(w, d, t, job, true)
 	case model.ActionAddTablePartition:
 		ver, err = rollingbackAddTablePartition(d, t, job)
+	case model.ActionReorganizePartition:
+		ver, err = rollingbackReorganizePartition(d, t, job)
 	case model.ActionDropColumn:
 		ver, err = rollingbackDropColumn(d, t, job)
 	case model.ActionDropIndex, model.ActionDropPrimaryKey:

